framework/src/node/consensus/certificate_generation/commit_pool.ts

```diff
@@ -372,10 +384,160 @@ export class CommitPool {
 		};
 	}
 
+	private async _job(apiContext: APIContext): Promise<void> {
+		const removalHeight = await this._getMaxRemovalHeight();
+		const { maxHeightPrecommitted } = await this._bftAPI.getBFTHeights(apiContext);
+
+		// 1. Clean up nonGossipedCommits
+		for (const height of this._nonGossipedCommits.keys()) {
+			// 1.1 Remove any single commit message m from nonGossipedCommits
+			if (height <= removalHeight) {
+				this._nonGossipedCommits.delete(height);
+				continue;
+			}
+			// 1.2 For every commit message m in nonGossipedCommits or gossipedCommits one of the following two conditions has to hold, otherwise it is discarded
+			const nonGossipedCommits = this._nonGossipedCommits.get(height) as SingleCommit[];
+			for (const singleCommit of nonGossipedCommits) {
+				// Condition #1
+				if (
+					!(
+						maxHeightPrecommitted - COMMIT_RANGE_STORED <= singleCommit.height ||
+						singleCommit.height <= maxHeightPrecommitted
+					)
+				) {
+					this._nonGossipedCommits.delete(singleCommit.height);
+					continue;
+				}
+				// Condition #2
+				const changeOfBFTParams = await this._bftAPI.existBFTParameters(
+					apiContext,
+					singleCommit.height + 1,
+				);
+				if (!changeOfBFTParams) {
+					this._nonGossipedCommits.delete(singleCommit.height);
+				}
+			}
+		}
+
+		// 1. Clean up gossipedCommits
+		for (const height of this._gossipedCommits.keys()) {
+			// 1.1 Remove any single commit message m from nonGossipedCommits
+			if (height <= removalHeight) {
+				this._gossipedCommits.delete(height);
+				continue;
+			}
+
+			// 1.2 For every commit message m in nonGossipedCommits or gossipedCommits one of the following two conditions has to hold, otherwise it is discarded
+			const gossipedCommits = this._gossipedCommits.get(height) as SingleCommit[];
+			for (const singleCommit of gossipedCommits) {
+				// Condition #1
+				if (
+					!(
+						maxHeightPrecommitted - COMMIT_RANGE_STORED < singleCommit.height ||
+						singleCommit.height < maxHeightPrecommitted
+					)
+				) {
+					this._gossipedCommits.delete(singleCommit.height);
+					continue;
+				}
+				// Condition #2
+				const changeOfBFTParams = await this._bftAPI.existBFTParameters(
+					apiContext,
+					singleCommit.height + 1,
+				);
+				if (!changeOfBFTParams) {
+					this._gossipedCommits.delete(singleCommit.height);
+				}
+			}
+		}
+
+		// 2. Select commits to gossip
+		const validators = await this._bftAPI.getCurrentValidators(apiContext);
+		const numActiveValidators = validators.length;
+		// Get a list of commits sorted by ascending order of height
+		const allCommits = this._getAllCommits();
+
+		const selectedCommits = [];
+		for (const commit of allCommits) {
+			if (selectedCommits.length >= numActiveValidators) {
+				break;
+			}
+
+			// 2.1 Choosing the commit with higher height first
+			if (commit.height < maxHeightPrecommitted - COMMIT_RANGE_STORED) {
```

Comment:
maybe this too ? but not sure

Reply:
all these conditions are checked in the same unit test created for above one